lib/lrmd/lrmd_client.c

```diff
@@ -1452,12 +1498,31 @@ lrmd_tcp_connect_cb(void *userdata, int rc, int sock)
         return;
     }
 
-    if (tls_client_handshake(lrmd) != pcmk_rc_ok) {
-        report_async_connection_result(lrmd, -EKEYREJECTED);
-        return;
-    }
+    /* If the TLS handshake immediately succeeds or fails, we can handle that
+     * now without having to deal with mainloops and retries.  Otherwise, add a
+     * timer event to keep trying until we get a result (or it times out).
+     */
+    rc = pcmk__tls_client_try_handshake(native->remote, &tls_rc);
+    if (rc == EAGAIN) {
+        struct handshake_data_s *hs = NULL;
 
-    tls_handshake_succeeded(lrmd);
+        if (native->handshake_timer != 0) {
+            return;
+        }
+
+        hs = pcmk__assert_alloc(1, sizeof(struct handshake_data_s));
+        hs->lrmd = lrmd;
+        hs->start_time = time(NULL);
+        hs->timeout_sec = TLS_HANDSHAKE_TIMEOUT;
+
+        native->handshake_timer = g_timeout_add(250, try_handshake_cb, hs);
```

Comment:
We do want a trigger rather than a timer, because we don't want to introduce any delays in the handshake, we just want to process other mainloop events if any are queued. The trigger can set itself again if the handshake is not complete. When it returns to the mainloop, the mainloop will process any higher-priority queued events first, then call the trigger again. If there are no higher-priority events, it's immediate.